File: etc/[BJ 2150] Strongly Connected Component /Tarjan.cpp
//
//  main.cpp
//  cpp
//
//  Created by 나지혜 on 2021/03/02.
//
#include <cstdio>
#include <vector>
#include <stack>
#include <algorithm>
using namespace std;
const int MAX = 100001;
int V, E, cnt, dfsn[MAX];
vector<int> adj[MAX];
bool finished[MAX];
stack<int> S;
vector<vector<int>> SCC;
int Scnum;

int DFS(int v)
{
    dfsn[v] = ++cnt;
    S.push(v);
    int parent = dfsn[v];
    for(int next:adj[v]){
        if(dfsn[next] == 0)
        {
            parent = min(parent, DFS(next));
        }
        else if(!finished[next]){
            parent = min(parent, dfsn[next]);
        }
    }
    
    if(parent == dfsn[v])
    {
        vector<int> new_scc;
        while(1)
        {
            int top = S.top();
            S.pop();
            new_scc.push_back(top);
            finished[top] = true;
            if(top == v) break;
        }
        sort(new_scc.begin(), new_scc.end());
        SCC.push_back(new_scc);
        Scnum+=1;
    }
    return parent;
}

int main(){
    scanf("%d %d", &V, &E);
    for (int i=0; i<E; i++){
        int A, B;
        scanf("%d %d", &A, &B);
        adj[A].push_back(B);
    }
    for (int i=1; i<=V; i++)
    {
        if(dfsn[i] == 0){
            DFS(i);
        }
    }
    sort(SCC.begin(), SCC.end());
    
    printf("%d\n", Scnum);
    for(auto& curr: SCC)
    {
        for(int nowcurr:curr)
            printf("%d ", nowcurr);
        puts("-1");
    }
}

File: etc/[BJ 2150] Strongly Connected Component /Tarjan.py
#타잔알고리즘 -> 메모리초과 
import sys
sys.setrecursionlimit(10**9)
graph = {}
dfsn = {}
finished = {}
stack = []
scc = []
index = 0
def init(v):
    graph[v] = []
    dfsn[v] = -1
    finished[v] = False
def DFS(v):
    global index
    index+=1
    dfsn[v] = index
    stack.append(v)
    #자신의 dfsn, 자식들의 결과나 dfsn 중 가장 작은 번호를 parent에 저장 
    parent = dfsn[v]
    for vertex in graph[v]:
        if(dfsn[vertex] == -1):#방문한 적이 없으면 
            parent = min(parent, DFS(vertex))
        else:
            if(finished[vertex] == False): #방문은 했으나 아직 SCC로 추출되지 않은 상태(역방향 간선?)
                parent = min(parent, dfsn[vertex])
    #자신과 자신의 자손들이 도달가능한 제일 높은 정점이 자신일 경우 SCC 추출
    if(parent == dfsn[v]):
        new_scc = []
        while(True):
            top = stack.pop()
            new_scc.append(top)
            finished[top] = True
            if(top == v):
                break
        new_scc.sort()
        scc.append(new_scc)
    return parent
if __name__ == "__main__":
    V, E = map(int, input().split())
    for i in range(1, V+1):
        init(i)
    for i in range(E):
        A, B = map(int, input().split())
        graph[A].append(B)
    
    for i in range(1, V+1):
        if(dfsn[i] == -1):
            DFS(i)
    
    scc.sort()
    print(len(scc))
    for i in scc:
        for k in i:
            print(k, end = " ")
        print(-1)
